### PR TITLE
Support tags for screenshots

### PR DESCRIPTION
**Pull request by solker**
Allow tagging screenshots #94

### CODE REVIEW

File: shoov/modules/custom/shoov_repository/shoov_repository.module [hunk 1]
```diff
@@ -37,7 +37,77 @@ function shoov_repository_node_presave($node) {
...
+/**
+ * Implements hook_node_insert().
+ *
+ * Check that vocabulary with name "screenshots_tags" exist for this repository
+ * otherwise create one and assign with repository.
+ *
+ * @param $node
+ *    Repository object.
+ */
+function shoov_repository_node_insert($node) {
```
Comment (solker) [inline shoov/modules/custom/shoov_repository/shoov_repository.module]: `if ($node->type != 'repository') {`

**Comment by solker** [comment]
@amitaibu done, ready for code review.

The problem I can't see widget to edit tags in display mode.
selection_118

**Comment by solker** [comment]
@amitaibu I have renamed the field.
selection_119

File: shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php [hunk 2]
```diff
@@ -88,5 +89,26 @@ public function complete($entity, $row) {
     $wrapper = entity_metadata_wrapper('node', $ui_build_id);
     $wrapper->field_pr_screenshot_ids->set($entity->nid);
     $wrapper->save();
+
+    $node = node_load($entity->nid);
+    $wrapper = entity_metadata_wrapper('node', $node);
+
+    // Check the vocabulary 'screenshots_tags' exist for this repository.
+    if (!$vocabulary_id = shoov_repository_vocabulary_exist('Screenshots tags', $wrapper->og_repo->value())) {
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php]: why is this needed?

File: shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php [hunk 3]
```diff
@@ -88,5 +89,26 @@ public function complete($entity, $row) {
     $wrapper = entity_metadata_wrapper('node', $ui_build_id);
     $wrapper->field_pr_screenshot_ids->set($entity->nid);
     $wrapper->save();
+
+    $node = node_load($entity->nid);
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php]: this shouldn't be in `complete` - it should happen in `prepare`

File: shoov/modules/custom/shoov_repository/shoov_repository.module [hunk 4]
```diff
@@ -37,7 +37,83 @@ function shoov_repository_node_presave($node) {
...
+ * @param $vocabulary
+ *    The name of a vocabulary.
+ * @param $node
+ *    The repository object or the id of repository.
+ *
+ * @return int|false
+ *    The vid of vocabulary or false if vocabulary doesn't exist.
+ */
+function shoov_repository_vocabulary_exist($vocabulary, $repository) {
+  if ($repository_id = is_int($repository)) {
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_repository/shoov_repository.module]: remove this checks

File: shoov/modules/custom/shoov_repository/shoov_repository.module [hunk 5]
```diff
@@ -37,7 +37,83 @@ function shoov_repository_node_presave($node) {
...
+    $repository = node_load($repository_id);
+    if (!$repository) {
+      throw Exception("The repository ID {$repository_id} doesn't exists");
+    }
+  }
+
+  // Check that the vocabulary exist for this repository.
+  $vocabularies = og_vocab_relation_get_by_group('node', $repository->nid);
+  foreach ($vocabularies as $voc) {
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_repository/shoov_repository.module]: voc => vocab

File: shoov/modules/custom/shoov_repository/shoov_repository.module [hunk 6]
```diff
@@ -37,7 +37,83 @@ function shoov_repository_node_presave($node) {
...
+    if (taxonomy_vocabulary_load($voc->vid)->name == $vocabulary) {
+      return $voc->vid;
+    }
+  }
+
+  return false;
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_repository/shoov_repository.module]: `return FALSE;`

File: shoov/modules/custom/shoov_repository/shoov_repository.module [hunk 7]
```diff
@@ -37,7 +37,83 @@ function shoov_repository_node_presave($node) {
...
+
+/**
+ * Determine the vocabulary exist in repository.
+ *
+ * @param $vocabulary
+ *    The name of a vocabulary.
+ * @param $node
+ *    The repository object or the id of repository.
+ *
+ * @return int|false
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_repository/shoov_repository.module]: `@return int|FALSE`

File: shoov/modules/custom/shoov_repository/shoov_repository.module [hunk 8]
```diff
@@ -37,7 +37,83 @@ function shoov_repository_node_presave($node) {
...
+  if (shoov_repository_vocabulary_exist('Screenshots tags', $node)) {
+    // Vocabulary for screenshots tags already exist.
+    return;
+  }
+
+  shoov_repository_create_screenshot_tags_vocabulary($node);
+}
+
+/**
+ * Determine the vocabulary exist in repository.
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_repository/shoov_repository.module]: `Determine if a vocabulary exists in a repository.`

File: shoov/modules/custom/shoov_repository/shoov_repository.module [hunk 9]
```diff
@@ -37,7 +37,83 @@ function shoov_repository_node_presave($node) {
...
+ *
+ * @param $vocabulary
+ *    The name of a vocabulary.
+ * @param $node
+ *    The repository object or the id of repository.
+ *
+ * @return int|false
+ *    The vid of vocabulary or false if vocabulary doesn't exist.
+ */
+function shoov_repository_vocabulary_exist($vocabulary, $repository) {
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_repository/shoov_repository.module]: `shoov_repository_get_vocabulary_by_repo`

File: shoov/modules/custom/shoov_repository/shoov_repository.module [hunk 10]
```diff
@@ -37,7 +37,83 @@ function shoov_repository_node_presave($node) {
...
+ *    The id of newly created vocabulary.
+ */
+function shoov_repository_create_screenshot_tags_vocabulary($node) {
+  // Creating the new vocabulary.
+  $vocabulary = new stdClass();
+  $vocabulary->name = "Screenshots tags";
+  $vocabulary->machine_name = "screenshots_tags_" . $node->nid;
+  $vocabulary->description = "The vocabulary contains tags for screenshots.";
+  taxonomy_vocabulary_save($vocabulary);
+  og_vocab_relation_save($vocabulary->vid, 'node', $node->nid);
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_repository/shoov_repository.module:131]: `// Attach the new vocabulary to the group.`

File: shoov/modules/custom/shoov_restful/plugins/restful/file/screenshots_upload/1.0/ShoovScreenshotsUploadResource.class.php [hunk 11]
```diff
@@ -59,6 +59,20 @@ protected function createScreenshotNode(array $files) {
...
+    if (!$vocabulary_id = shoov_repository_vocabulary_exist('screenshots_tags', $repo_node)) {
+      $vocabulary_id = shoov_repository_create_screenshot_tags_vocabulary($node);
+    }
+
+    if ($request['tags']) {
+      $tags = explode(',', $request['tags']);
+      $tids = array();
+      foreach($tags as $tag) {
+        $tid = shoov_screenshot_add_tag_to_vocabulary($tag, $vocabulary_id);
+        array_push($tids, $tid);
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_restful/plugins/restful/file/screenshots_upload/1.0/ShoovScreenshotsUploadResource.class.php]: `$tids[] = $tid;`

File: shoov/modules/custom/shoov_screenshot/shoov_screenshot.module [hunk 12]
```diff
@@ -82,3 +82,38 @@ function shoov_screenshot_stream_wrappers() {
   return $wrappers;
 }
 
+/**
+ * Function checks if tag already exist or creates new tag.
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_screenshot/shoov_screenshot.module]: `Adds an existing tag or create a new one and adds to a repository vocabulary.`

File: shoov/modules/custom/shoov_screenshot/shoov_screenshot.module [hunk 13]
```diff
@@ -82,3 +82,38 @@ function shoov_screenshot_stream_wrappers() {
...
+ * And it adds this tag to vocabulary.
+ *
+ * @param $tag
+ *    The name of tag
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_screenshot/shoov_screenshot.module]: missing dots

File: shoov/modules/custom/shoov_screenshot/shoov_screenshot.module [hunk 14]
```diff
@@ -82,3 +82,38 @@ function shoov_screenshot_stream_wrappers() {
...
+ * @param $vocabulary
+ *    The id of vocabulary
+ *
+ * @return int
+ *    The tid of tag
+ */
+function shoov_screenshot_add_tag_to_vocabulary($tag, $vocabulary) {
+  $query  = new EntityFieldQuery();
+
+  $results = $query
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_screenshot/shoov_screenshot.module]: results => result

File: shoov/modules/custom/shoov_screenshot/shoov_screenshot.module [hunk 15]
```diff
@@ -82,3 +82,38 @@ function shoov_screenshot_stream_wrappers() {
...
+ * Function checks if tag already exist or creates new tag.
+ * And it adds this tag to vocabulary.
+ *
+ * @param $tag
+ *    The name of tag
+ * @param $vocabulary
+ *    The id of vocabulary
+ *
+ * @return int
+ *    The tid of tag
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_screenshot/shoov_screenshot.module]: `The term ID.`

File: shoov/modules/custom/shoov_screenshot/shoov_screenshot.module [hunk 16]
```diff
@@ -82,3 +82,38 @@ function shoov_screenshot_stream_wrappers() {
...
+ */
+function shoov_screenshot_add_tag_to_vocabulary($tag, $vocabulary) {
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_screenshot/shoov_screenshot.module]: vocabulary => vocabulary_id

Comment (amitaibu) [inline shoov/modules/custom/shoov_screenshot/shoov_screenshot.module]: $tag => $term_name

File: shoov/modules/custom/shoov_screenshot/shoov_screenshot.module [hunk 17]
```diff
@@ -82,3 +82,38 @@ function shoov_screenshot_stream_wrappers() {
...
+  $query  = new EntityFieldQuery();
+
+  $results = $query
+    ->entityCondition('entity_type', 'taxonomy_term')
+    ->propertyCondition('name', $tag)
+    ->propertyCondition('vid', $vocabulary)
+    ->execute();
+
+  if (empty($results['taxonomy_term'])) {
+    $term = new stdClass();
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_screenshot/shoov_screenshot.module:117]: `// Create a new term.`

File: shoov/modules/custom/shoov_screenshot/shoov_screenshot.module [hunk 18]
```diff
@@ -82,3 +82,38 @@ function shoov_screenshot_stream_wrappers() {
...
+    $term->name = $tag;
+    $term->vid = $vocabulary;
+    taxonomy_term_save($term);
+    $tid = $term->tid;
+  }
+  else {
+    $keys = array_keys($results['taxonomy_term']);
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_screenshot/shoov_screenshot.module]: `$key = key($result['taxonomy_term']);`

File: shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php (hunk 2 above)
Comment (solker) [inline shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php]: To be sure that vocabulary exist. It should never happen.

File: shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php (hunk 3 above)
Comment (solker) [inline shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php]: I think on the `prepare` state a vocabulary still not exist because the vocabulary is creating in `hook_node_insert` hook.

File: shoov/modules/custom/shoov_repository/shoov_repository.module (hunk 4 above)
Comment (solker) [inline shoov/modules/custom/shoov_repository/shoov_repository.module]: It's not really check. This part of code allow you to use ID of Repository instead of repository object.

File: shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php (hunk 3 above)
Comment (amitaibu) [inline shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php]: it's created for the repo - not for the screenshot

File: shoov/modules/custom/shoov_repository/shoov_repository.module (hunk 10 above)
Comment (amitaibu) [inline shoov/modules/custom/shoov_repository/shoov_repository.module:131]: need to add comment^^

**Comment by amitaibu** [comment]
Needs a re-roll

File: shoov/modules/custom/shoov_repository/shoov_repository.module [hunk 19]
```diff
@@ -45,6 +46,81 @@ function shoov_repository_node_presave($node) {
 }
 
 /**
+<<<<<<< HEAD
```
Comment (solker) [inline shoov/modules/custom/shoov_repository/shoov_repository.module]: @amitaibu need help to solve the conflict.

File: shoov/modules/custom/shoov_restful/plugins/restful/file/screenshots_upload/1.0/ShoovScreenshotsUploadResource.class.php [hunk 20]
```diff
@@ -59,6 +59,20 @@ protected function createScreenshotNode(array $files) {
 
     $wrapper->field_baseline_name->set($request['baseline_name']);
 
+    if (!$vocabulary_id = shoov_repository_get_vocabulary_by_repo('screenshots_tags', $repo_node)) {
+      $vocabulary_id = shoov_repository_create_screenshot_tags_vocabulary($node);
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_restful/plugins/restful/file/screenshots_upload/1.0/ShoovScreenshotsUploadResource.class.php]: remove, assume it is there. please open a follow up issue to write script to update existing repos

File: shoov/modules/custom/shoov_screenshot/shoov_screenshot.info [hunk 21]
```diff
@@ -19,14 +20,19 @@ features[field_base][] = field_build
...
 features[field_base][] = field_regression_image
 features[field_base][] = og_repo
+features[field_base][] = og_vocabulary
 features[field_instance][] = node-screenshot-field_baseline_image
 features[field_instance][] = node-screenshot-field_baseline_name
 features[field_instance][] = node-screenshot-field_build
 features[field_instance][] = node-screenshot-field_diff_image
 features[field_instance][] = node-screenshot-field_regression_image
 features[field_instance][] = node-screenshot-og_repo
+<<<<<<< HEAD
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_screenshot/shoov_screenshot.info]: conflict

**Comment by solker** [comment]
@amitaibu Travis is green, probably ready for code review.

File: shoov/modules/custom/shoov_repository/shoov_repository.module [hunk 22]
```diff
@@ -45,6 +46,102 @@ function shoov_repository_node_presave($node) {
...
+ *
+ * @param $vocabulary
+ *    The name of a vocabulary.
+ * @param $node
+ *    The repository object or the id of repository.
+ *
+ * @return int|FALSE
+ *    The vid of vocabulary or false if vocabulary doesn't exist.
+ */
+function shoov_repository_get_vocabulary_by_repo($vocabulary, $repository) {
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_repository/shoov_repository.module]: lets force sending just the node ID, not the whole object.

File: shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php [hunk 23]
```diff
@@ -89,4 +90,32 @@ public function complete($entity, $row) {
...
+
+    // If a screenshot have tags handle it.
+    if ($row->_tags) {
+      $tags = explode(',', $row->_tags);
+      $tids = array();
+      foreach($tags as $tag) {
+        $tid = shoov_screenshot_add_tag_to_vocabulary($tag, $vocabulary_id);
+        $tids[] = $tid;
+      }
+      $wrapper->og_vocabulary->set($tids);
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php]: `$wrapper->og_vocabulary->set` => `$wrapper->{OG_VOCAB_FIELD}->set`

File: shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php [hunk 24]
```diff
@@ -89,4 +90,32 @@ public function complete($entity, $row) {
...
+    }
+
+    $wrapper->save();
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php]: remove

File: shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php [hunk 25]
```diff
@@ -89,4 +90,32 @@ public function complete($entity, $row) {
...
+   *
+   * Handle tags.
+   *
+   * @param $entity
+   * @param $row
+   */
+  function prepare($entity, $row) {
+    $wrapper = entity_metadata_wrapper('node', $entity);
+
+    // Check the vocabulary 'screenshots_tags' exist for this repository.
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php]: remove comment, as it's wrong

File: shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php [hunk 26]
```diff
@@ -89,4 +90,32 @@ public function complete($entity, $row) {
     $wrapper->field_pr_screenshot_ids->set($entity->nid);
     $wrapper->save();
   }
+
+  /**
+   * Implements MigrateDestination::prepare().
+   *
+   * Handle tags.
+   *
+   * @param $entity
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php]: remove params

File: shoov/modules/custom/shoov_restful/plugins/restful/file/screenshots_upload/1.0/ShoovScreenshotsUploadResource.class.php [hunk 27]
```diff
@@ -59,6 +59,18 @@ protected function createScreenshotNode(array $files) {
...
+    $vocabulary_id = shoov_repository_get_vocabulary_by_repo('screenshots_tags', $repo_node);
+
+    if ($request['tags']) {
+      $tags = explode(',', $request['tags']);
+      $tids = array();
+      foreach($tags as $tag) {
+        $tid = shoov_screenshot_add_tag_to_vocabulary($tag, $vocabulary_id);
+        $tids[] = $tid;
+      }
+      $wrapper->og_vocabulary->set($tids);
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_restful/plugins/restful/file/screenshots_upload/1.0/ShoovScreenshotsUploadResource.class.php]: => `$wrapper->{OG_VOCAB_FIELD}`

File: shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php (hunk 23 above)
Comment (solker) [inline shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php]: What is different?

Comment (solker) [inline shoov/modules/custom/shoov_migrate/handlers/node/ShoovScreenshotsMigrate.php]: Ah, understood. Sorry.

**Comment by solker** [comment]
Hm, by logic I should have constant `OG_REPO_FIELD`.
But I don't.

``` php
$wrapper->{OG_VOCAB_FIELD}->set($tids);
$wrapper->og_repo->set($repo_node);
```

**Comment by solker** [comment]
@amitaibu ok, now is green.

File: shoov/modules/custom/shoov_repository/shoov_repository.module [hunk 28]
```diff
@@ -45,6 +46,95 @@ function shoov_repository_node_presave($node) {
...
+ * @param $node
+ *    The repository id.
+ *
+ * @return int
+ *    The id of newly created vocabulary.
+ */
+function shoov_repository_create_screenshot_tags_vocabulary($node) {
+  if (shoov_repository_get_vocabulary_by_repo('Screenshots tags', $node->nid)) {
+    // Vocabulary for screenshots tags already exist.
+    return;
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_repository/shoov_repository.module]: maybe the existing vocab ID.

**Comment by amitaibu** [comment]
Merged, thanks.
